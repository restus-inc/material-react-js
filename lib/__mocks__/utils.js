/* global jest */
const utils = jest.genMockFromModule('../utils');

const newId = () => 'a0123456';
utils.newId = newId;

export default utils;
export { newId };
