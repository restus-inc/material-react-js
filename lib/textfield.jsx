import React, { useEffect, useRef } from 'react';
import { MDCTextField } from '@material/textfield';

const NON_NATIVE_PROPS = [
  'id', 'value', 'defaultValue', 'type', 'variation', 'label', 'className', 'helperText',
  'showsHelperPersistently', 'showsHelperAsValidation', 'rows', 'cols', 'resizable',
];

const generateRootClassName = (props) => {
  const rootClasses = ['mdc-text-field'];
  if (props.label == null || props.label === '') {
    rootClasses.push('mdc-text-field--no-label');
  }
  if (props.disabled) {
    rootClasses.push('mdc-text-field--disabled');
  }
  if (props.defaultValue) {
    rootClasses.push('mdc-text-field--label-floating');
  }

  switch (props.variation || 'filled') {
    case 'filled':
      rootClasses.push('mdc-text-field--filled');
      break;
    case 'outlined':
      rootClasses.push('mdc-text-field--outlined');
      break;
    case 'textarea':
      rootClasses.push('mdc-text-field--outlined', 'mdc-text-field--textarea');
      break;
    case 'filled-textarea':
      rootClasses.push('mdc-text-field--filled', 'mdc-text-field--textarea');
      break;
    default:
      throw new Error(`Not suported variation; ${props.variation}`);
  }
  if (props.className) {
    rootClasses.push(props.className);
  }
  return rootClasses.join(' ');
};

function NativeInput(props) {
  const nativeInputProps = Object.entries(props).reduce((newProps, [key, value]) => (
    !NON_NATIVE_PROPS.includes(key) ? Object.assign(newProps, { [key]: value }) : newProps
  ), { className: 'mdc-text-field__input' });
  if (props.value == null) {
    Object.assign(nativeInputProps, { defaultValue: props.defaultValue });
  } else {
    Object.assign(nativeInputProps, { value: props.value });
  }
  if (props.dataset) {
    Object.assign(
      nativeInputProps,
      ...Object.entries(props.dataset).map(([key, value]) => ({ [`data-${key}`]: value })),
    );
  }
  if ((props.variation || 'filled').endsWith('textarea')) {
    Object.assign(nativeInputProps, { rows: props.rows || 4, cols: props.cols || 32 });
  } else {
    Object.assign(nativeInputProps, { type: props.type || 'text' });
  }
  if (props.id) {
    Object.assign(nativeInputProps, { id: props.id, name: props.id });
    if (!(props.label == null || props.label === '')) {
      Object.assign(nativeInputProps, { 'aria-labelledby': `${props.id}-label` });
    }
    if (props.helperText || props.showsHelperPersistently || props.showsHelperAsValidation) {
      Object.assign(nativeInputProps, {
        'aria-controls': `${props.id}-helper`,
        'aria-describedby': `${props.id}-helper`,
      });
    }
  }
  const tagName = (props.variation || 'filled').endsWith('textarea') ? 'textarea' : 'input';
  return React.createElement(tagName, nativeInputProps);
}

function HelperText(props) {
  if (!props.helperText && !props.showsHelperPersistently && !props.showsHelperAsValidation) {
    return null;
  }
  const classList = ['mdc-text-field-helper-text'];
  if (props.showsHelperPersistently) {
    classList.push('mdc-text-field-helper-text--persistent');
  }
  if (props.showsHelperAsValidation) {
    classList.push('mdc-text-field-helper-text--validation-msg');
  }
  return (
    <div className="mdc-text-field-helper-line">
      <div className={classList.join(' ')} id={props.id && `${props.id}-helper`} aria-hidden="true">{props.helperText}</div>
    </div>
  );
}

/**
 * [MDCTextField component]{@link https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield#readme}
 * implemented by react component.
 * @function TextField
 * @param {Object} [props] Attributes other than followings are passed to the `input`
 * or `textarea` element of React as is.
 * @param {string} [props.id] The `id` attribute of the native html input element. This
 * property is also used the `name` attribute.
 * @param {string} [props.value] The value of the text field. Should not specify both
 * this attribute and `props.defaultValue` at the same time.
 * @param {string} [props.defaultValue] The default value of the text field. Should not
 * specify both this attribute and `props.value` at the same time.
 * @param {string} [props.type] The `type` attribute of the native html input element.
 * If `props.variation` is `'textarea'` or `'filled-textarea'`, this attribute is ignored.
 * @param {string} [props.variation] The variation of the text field. Supported variations
 *  are`'filled'`(default), `'outlined'`, `'textarea'` and `'filled-textarea'`.
 * @param {string} [props.label] The label text of the text field.
 * @param {string} [props.className] The class name that is added to the root element.
 * @param {boolean} [props.disabled] Specifies `true` if you want to disable the text field.
 * Default to `false`.
 * @param {string} [props.helperText] Specifeis the helper text.
 * @param {boolean} [props.showsHelperPersistently] Specifies `true` to make the helper text
 * permanently visible. Default to `false`.
 * @param {boolean} [props.showsHelperAsValidation] Specifies `true` if the helper text is
 * a validation message. Default to `false`.
 * @param {number} [props.rows] The number of rows of textarea. If `props.variation` is not
 * `'textarea'` or `'filled-textarea'`, this attribute is ignored. Default to `4`.
 * @param {number} [props.cols] The cols attribute of textarea element. If `props.variation`
 * is not `'textarea'` or `'filled-textarea'`, this attribute is ignored. Default to `32`.
 * @param {boolean} [props.resizable] Specifies `true` allowing the user to resize the textarea.
 * If `props.variation` is not `'textarea'` or `'filled-textarea'`, this attribute is ignored.
 * Default to `false`
 * @returns {DetailedReactHTMLElement}
 * @exports material-react-js
 */
export default function TextField(props) {
  const rootElement = useRef();
  const mdcComponentRef = useRef();

  useEffect(() => {
    const hasFocus = document.hasFocus() && document.activeElement === rootElement.current;
    if (mdcComponentRef.current && !hasFocus) {
      mdcComponentRef.current.useNativeValidation = false;
      mdcComponentRef.current.value = props.value;
      mdcComponentRef.current.useNativeValidation = true;
    }
  }, [props.value]);

  useEffect(() => {
    mdcComponentRef.current = new MDCTextField(rootElement.current);
    return () => {
      mdcComponentRef.current.destroy();
    };
  }, []);

  const rootClassName = generateRootClassName(props);
  const labelClassName = props.defaultValue
    ? 'mdc-floating-label mdc-floating-label--float-above'
    : 'mdc-floating-label';

  switch (props.variation || 'filled') {
    case 'filled':
      return (
        <>
          <label className={rootClassName} ref={rootElement}>
            <span className="mdc-text-field__ripple"></span>
            <NativeInput {...props}/>
            {!(props.label == null || props.label === '') && (
            <span className={labelClassName} id={props.id && `${props.id}-label`}>{props.label}</span>
            )}
            <span className="mdc-line-ripple"></span>
          </label>
          <HelperText {...props}/>
        </>
      );
    case 'outlined':
      return (
        <>
          <label className={rootClassName} ref={rootElement}>
          <NativeInput {...props}/>
            <span className="mdc-notched-outline">
              <span className="mdc-notched-outline__leading"></span>
              {!(props.label == null || props.label === '') && (
                <span className="mdc-notched-outline__notch">
                  <span className={labelClassName} id={props.id && `${props.id}-label`}>{props.label}</span>
                </span>
              )}
              <span className="mdc-notched-outline__trailing"></span>
            </span>
          </label>
          <HelperText {...props}/>
        </>
      );
    case 'textarea':
      return (
        <>
          <label className={rootClassName} ref={rootElement}>
            {props.resizable ? (
              <span className="mdc-text-field__resizer">
                <NativeInput {...props}/>
              </span>
            ) : (
              <NativeInput {...props}/>
            )}
            <span className="mdc-notched-outline">
              <span className="mdc-notched-outline__leading"></span>
              {!(props.label == null || props.label === '') && (
                <span className="mdc-notched-outline__notch">
                  <span className={labelClassName} id={props.id && `${props.id}-label`}>{props.label}</span>
                </span>
              )}
              <span className="mdc-notched-outline__trailing"></span>
            </span>
          </label>
          <HelperText {...props}/>
        </>
      );
    case 'filled-textarea':
      return (
        <>
          <label className={rootClassName} ref={rootElement}>
            <span className="mdc-text-field__ripple"></span>
            {props.resizable ? (
              <span className="mdc-text-field__resizer">
                <NativeInput {...props}/>
              </span>
            ) : (
              <NativeInput {...props}/>
            )}
            {!(props.label == null || props.label === '') && (
            <span className={labelClassName} id={props.id && `${props.id}-label`}>{props.label}</span>
            )}
            <span className="mdc-line-ripple"></span>
          </label>
          <HelperText {...props}/>
        </>
      );
    default:
      throw new Error(`Not suported variation; ${props.variation}`);
  }
}
