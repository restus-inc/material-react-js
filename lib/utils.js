/**
 * Utilities for Material React.
 * @module material-react/lib/utils
 */

/**
 * Generates a rundom ID. The generated ID starts with basic latin child alphabet, and
 * subsequent characters consist of child alphanumeric character. (i.e. it matches
 * `/[a-z][a-z0-9]{digit-1}/`)
 * @param {number} [digit=8] The number of digit of the generated ID
 * @returns {string} The ID generated
 */
export function newId(digit = 8) {
  return (
    (Math.floor(Math.random() * 26) + 10) * (36 ** (digit - 1))
    + Math.floor(Math.random() * (36 ** (digit - 1)))
  ).toString(36);
}

/**
 * Converts the event name to event handler name.
 * @param {string} eventName The event name that is converted to the handler name
 * @return {string} The event handler name converted
 */
export function handlerName(eventName) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
