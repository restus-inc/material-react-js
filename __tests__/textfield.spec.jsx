/**
 * @license
 * Copyright 2020 Restus Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* global describe, afterEach, it, expect */
import 'regenerator-runtime/runtime';
import React, { useEffect, useRef } from 'react';
import { MDCTextField } from '@material/textfield';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { htmlOfRendering } from './utils';

import TextField from '../lib/textfield';

describe('TextField component', () => {
  afterEach(async () => {
    await cleanup();
  });

  it('supports filled text fields', () => {
    expect(htmlOfRendering(
      <TextField/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField label="foo" placeholder="bar" required={true} disabled={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField label="foo" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="filled" label="foo" id="bar" defaultValue="baz"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField helperText="qux"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField label="foo" placeholder="bar" required={true} disabled={true} helperText="qux" showsHelperPersistently={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField label="foo" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*" helperText="qux" showsHelperAsValidation={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('supports outlined text fields', () => {
    expect(htmlOfRendering(
      <TextField variation="outlined"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="outlined" label="foo" placeholder="bar" required={true} disabled={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="outlined" label="foo" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="outlined" helperText="qux"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="outlined" label="foo" placeholder="bar" required={true} disabled={true} helperText="qux" showsHelperPersistently={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="outlined" label="foo" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*" helperText="qux" showsHelperAsValidation={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('supports outlined textareas', () => {
    expect(htmlOfRendering(
      <TextField variation="textarea"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="textarea" resizable={true} placeholder="bar" required={true} disabled={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="textarea" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*" rows={6} cols={40}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="textarea" resizable={true} helperText="qux"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="textarea" label="foo" placeholder="bar" required={true} disabled={true} helperText="qux" showsHelperPersistently={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="textarea" label="foo" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*" helperText="qux" showsHelperAsValidation={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('supports filled textareas', () => {
    expect(htmlOfRendering(
      <TextField variation="filled-textarea"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="filled-textarea" resizable={true} placeholder="bar" required={true} disabled={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="filled-textarea" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*" rows={6} cols={40}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="filled-textarea" resizable={true} helperText="qux"/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="filled-textarea" label="foo" placeholder="bar" required={true} disabled={true} helperText="qux" showsHelperPersistently={true}/>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfRendering(
      <TextField variation="filled-textarea" label="foo" id="bar" defaultValue="baz" className="qux" pattern="[a-z]*" helperText="qux" showsHelperAsValidation={true}/>,
    )).resolves.toMatchSnapshot();
  });

  it('can provide an MDCTextField instance', async () => {
    let mdcTextFieldComponent;
    function MyTextField() {
      const mdcTextFieldRef = useRef();
      useEffect(() => {
        mdcTextFieldComponent = mdcTextFieldRef.current;
      });
      return <TextField label="foo" mdcTextFieldRef={mdcTextFieldRef}/>;
    }
    render(<MyTextField/>);
    expect(mdcTextFieldComponent).toBeInstanceOf(MDCTextField);
  });

  it('can be manipulated via MDCTextField', async () => {
    function TextfieldTester() {
      const mdcTextFieldRef = useRef();
      function changeText() {
        mdcTextFieldRef.current.value = 'baz qux';
      }
      return (
         <>
          <TextField id="my-text-field" label="text" defaultValue="foo bar" mdcTextFieldRef={mdcTextFieldRef} data-testid="my-text-field"/>
          <button type="button" onClick={changeText}>change</button>
        </>
      );
    }
    const { getByTestId, getByText } = render(<TextfieldTester/>);
    const nativeInput = getByTestId('my-text-field');
    const button = getByText('change');

    expect(nativeInput.value).toBe('foo bar');
    userEvent.click(button);
    expect(nativeInput.value).toBe('baz qux');
  });
});
