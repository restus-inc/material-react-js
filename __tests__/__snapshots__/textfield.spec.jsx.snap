// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField component supports filled text fields 1`] = `
<label
  className="mdc-text-field mdc-text-field--no-label mdc-text-field--filled"
>
  <span
    className="mdc-text-field__ripple"
  />
  <input
    className="mdc-text-field__input"
    disabled={false}
    required={false}
    type="text"
  />
  <span
    className="mdc-line-ripple"
  />
</label>
`;

exports[`TextField component supports filled text fields 2`] = `
<label
  className="mdc-text-field mdc-text-field--disabled mdc-text-field--filled"
>
  <span
    className="mdc-text-field__ripple"
  />
  <input
    className="mdc-text-field__input"
    disabled={true}
    placeholder="bar"
    required={false}
    type="text"
  />
  <span
    className="mdc-floating-label"
  >
    foo
  </span>
  <span
    className="mdc-line-ripple"
  />
</label>
`;

exports[`TextField component supports filled text fields 3`] = `
<label
  className="mdc-text-field mdc-text-field--label-floating mdc-text-field--filled qux"
>
  <span
    className="mdc-text-field__ripple"
  />
  <input
    aria-labelledby="bar-label"
    className="mdc-text-field__input"
    defaultValue="baz"
    disabled={false}
    id="bar"
    name="bar"
    pattern="[a-z]*"
    required={false}
    type="text"
  />
  <span
    className="mdc-floating-label mdc-floating-label--float-above"
    id="bar-label"
  >
    foo
  </span>
  <span
    className="mdc-line-ripple"
  />
</label>
`;

exports[`TextField component supports filled text fields 4`] = `
<label
  className="mdc-text-field mdc-text-field--label-floating mdc-text-field--filled"
>
  <span
    className="mdc-text-field__ripple"
  />
  <input
    aria-labelledby="bar-label"
    className="mdc-text-field__input"
    defaultValue="baz"
    disabled={false}
    id="bar"
    name="bar"
    required={false}
    type="text"
  />
  <span
    className="mdc-floating-label mdc-floating-label--float-above"
    id="bar-label"
  >
    foo
  </span>
  <span
    className="mdc-line-ripple"
  />
</label>
`;

exports[`TextField component supports filled text fields 5`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--no-label mdc-text-field--filled"
  >
    <span
      className="mdc-text-field__ripple"
    />
    <input
      className="mdc-text-field__input"
      disabled={false}
      required={false}
      type="text"
    />
    <span
      className="mdc-line-ripple"
    />
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports filled text fields 6`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--disabled mdc-text-field--filled"
  >
    <span
      className="mdc-text-field__ripple"
    />
    <input
      className="mdc-text-field__input"
      disabled={true}
      placeholder="bar"
      required={false}
      type="text"
    />
    <span
      className="mdc-floating-label"
    >
      foo
    </span>
    <span
      className="mdc-line-ripple"
    />
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text mdc-text-field-helper-text--persistent"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports filled text fields 7`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--label-floating mdc-text-field--filled qux"
  >
    <span
      className="mdc-text-field__ripple"
    />
    <input
      aria-controls="bar-helper"
      aria-describedby="bar-helper"
      aria-labelledby="bar-label"
      className="mdc-text-field__input"
      defaultValue="baz"
      disabled={false}
      id="bar"
      name="bar"
      pattern="[a-z]*"
      required={false}
      type="text"
    />
    <span
      className="mdc-floating-label mdc-floating-label--float-above"
      id="bar-label"
    >
      foo
    </span>
    <span
      className="mdc-line-ripple"
    />
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text mdc-text-field-helper-text--validation-msg"
      id="bar-helper"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports outlined text fields 1`] = `
<label
  className="mdc-text-field mdc-text-field--no-label mdc-text-field--outlined"
>
  <input
    className="mdc-text-field__input"
    disabled={false}
    required={false}
    type="text"
  />
  <span
    className="mdc-notched-outline"
  >
    <span
      className="mdc-notched-outline__leading"
    />
    <span
      className="mdc-notched-outline__trailing"
    />
  </span>
</label>
`;

exports[`TextField component supports outlined text fields 2`] = `
<label
  className="mdc-text-field mdc-text-field--disabled mdc-text-field--outlined"
>
  <input
    className="mdc-text-field__input"
    disabled={true}
    placeholder="bar"
    required={false}
    type="text"
  />
  <span
    className="mdc-notched-outline"
  >
    <span
      className="mdc-notched-outline__leading"
    />
    <span
      className="mdc-notched-outline__notch"
    >
      <span
        className="mdc-floating-label"
      >
        foo
      </span>
    </span>
    <span
      className="mdc-notched-outline__trailing"
    />
  </span>
</label>
`;

exports[`TextField component supports outlined text fields 3`] = `
<label
  className="mdc-text-field mdc-text-field--label-floating mdc-text-field--outlined qux"
>
  <input
    aria-labelledby="bar-label"
    className="mdc-text-field__input"
    defaultValue="baz"
    disabled={false}
    id="bar"
    name="bar"
    pattern="[a-z]*"
    required={false}
    type="text"
  />
  <span
    className="mdc-notched-outline"
  >
    <span
      className="mdc-notched-outline__leading"
    />
    <span
      className="mdc-notched-outline__notch"
    >
      <span
        className="mdc-floating-label mdc-floating-label--float-above"
        id="bar-label"
      >
        foo
      </span>
    </span>
    <span
      className="mdc-notched-outline__trailing"
    />
  </span>
</label>
`;

exports[`TextField component supports outlined text fields 4`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--no-label mdc-text-field--outlined"
  >
    <input
      className="mdc-text-field__input"
      disabled={false}
      required={false}
      type="text"
    />
    <span
      className="mdc-notched-outline"
    >
      <span
        className="mdc-notched-outline__leading"
      />
      <span
        className="mdc-notched-outline__trailing"
      />
    </span>
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports outlined text fields 5`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--disabled mdc-text-field--outlined"
  >
    <input
      className="mdc-text-field__input"
      disabled={true}
      placeholder="bar"
      required={false}
      type="text"
    />
    <span
      className="mdc-notched-outline"
    >
      <span
        className="mdc-notched-outline__leading"
      />
      <span
        className="mdc-notched-outline__notch"
      >
        <span
          className="mdc-floating-label"
        >
          foo
        </span>
      </span>
      <span
        className="mdc-notched-outline__trailing"
      />
    </span>
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text mdc-text-field-helper-text--persistent"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports outlined text fields 6`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--label-floating mdc-text-field--outlined qux"
  >
    <input
      aria-controls="bar-helper"
      aria-describedby="bar-helper"
      aria-labelledby="bar-label"
      className="mdc-text-field__input"
      defaultValue="baz"
      disabled={false}
      id="bar"
      name="bar"
      pattern="[a-z]*"
      required={false}
      type="text"
    />
    <span
      className="mdc-notched-outline"
    >
      <span
        className="mdc-notched-outline__leading"
      />
      <span
        className="mdc-notched-outline__notch"
      >
        <span
          className="mdc-floating-label mdc-floating-label--float-above"
          id="bar-label"
        >
          foo
        </span>
      </span>
      <span
        className="mdc-notched-outline__trailing"
      />
    </span>
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text mdc-text-field-helper-text--validation-msg"
      id="bar-helper"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports outlined textareas 1`] = `
<label
  className="mdc-text-field mdc-text-field--no-label mdc-text-field--outlined mdc-text-field--textarea"
>
  <span
    className="mdc-text-field__resizer"
  >
    <textarea
      className="mdc-text-field__input"
      cols={32}
      disabled={false}
      required={false}
      rows={4}
    />
  </span>
  <span
    className="mdc-notched-outline"
  >
    <span
      className="mdc-notched-outline__leading"
    />
    <span
      className="mdc-notched-outline__trailing"
    />
  </span>
</label>
`;

exports[`TextField component supports outlined textareas 2`] = `
<label
  className="mdc-text-field mdc-text-field--no-label mdc-text-field--disabled mdc-text-field--outlined mdc-text-field--textarea"
>
  <span
    className="mdc-text-field__resizer"
  >
    <textarea
      className="mdc-text-field__input"
      cols={32}
      disabled={true}
      placeholder="bar"
      required={false}
      rows={4}
    />
  </span>
  <span
    className="mdc-notched-outline"
  >
    <span
      className="mdc-notched-outline__leading"
    />
    <span
      className="mdc-notched-outline__trailing"
    />
  </span>
</label>
`;

exports[`TextField component supports outlined textareas 3`] = `
<label
  className="mdc-text-field mdc-text-field--no-label mdc-text-field--label-floating mdc-text-field--outlined mdc-text-field--textarea qux"
>
  <span
    className="mdc-text-field__resizer"
  >
    <textarea
      className="mdc-text-field__input"
      cols={40}
      defaultValue="baz"
      disabled={false}
      id="bar"
      name="bar"
      pattern="[a-z]*"
      required={false}
      rows={6}
    />
  </span>
  <span
    className="mdc-notched-outline"
  >
    <span
      className="mdc-notched-outline__leading"
    />
    <span
      className="mdc-notched-outline__trailing"
    />
  </span>
</label>
`;

exports[`TextField component supports outlined textareas 4`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--no-label mdc-text-field--outlined mdc-text-field--textarea"
  >
    <span
      className="mdc-text-field__resizer"
    >
      <textarea
        className="mdc-text-field__input"
        cols={32}
        disabled={false}
        required={false}
        rows={4}
      />
    </span>
    <span
      className="mdc-notched-outline"
    >
      <span
        className="mdc-notched-outline__leading"
      />
      <span
        className="mdc-notched-outline__trailing"
      />
    </span>
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports outlined textareas 5`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--disabled mdc-text-field--outlined mdc-text-field--textarea"
  >
    <span
      className="mdc-text-field__resizer"
    >
      <textarea
        className="mdc-text-field__input"
        cols={32}
        disabled={true}
        placeholder="bar"
        required={false}
        rows={4}
      />
    </span>
    <span
      className="mdc-notched-outline"
    >
      <span
        className="mdc-notched-outline__leading"
      />
      <span
        className="mdc-notched-outline__trailing"
      />
    </span>
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text mdc-text-field-helper-text--persistent"
    >
      qux
    </div>
  </div>,
]
`;

exports[`TextField component supports outlined textareas 6`] = `
Array [
  <label
    className="mdc-text-field mdc-text-field--label-floating mdc-text-field--outlined mdc-text-field--textarea qux"
  >
    <span
      className="mdc-text-field__resizer"
    >
      <textarea
        aria-controls="bar-helper"
        aria-describedby="bar-helper"
        aria-labelledby="bar-label"
        className="mdc-text-field__input"
        cols={32}
        defaultValue="baz"
        disabled={false}
        id="bar"
        name="bar"
        pattern="[a-z]*"
        required={false}
        rows={4}
      />
    </span>
    <span
      className="mdc-notched-outline"
    >
      <span
        className="mdc-notched-outline__leading"
      />
      <span
        className="mdc-notched-outline__trailing"
      />
    </span>
  </label>,
  <div
    className="mdc-text-field-helper-line"
  >
    <div
      aria-hidden="true"
      className="mdc-text-field-helper-text mdc-text-field-helper-text--validation-msg"
      id="bar-helper"
    >
      qux
    </div>
  </div>,
]
`;
